test/hotspot/jtreg/serviceability/jvmti/vthread/StopThreadTest/StopThreadTest.java

```diff
@@ -0,0 +1,242 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ *
+ * @summary Verifies JVMTI StopThread support for virtual threads.
+ * DESCRIPTION
+ *     The test exercises the JVMTI function: StopThread(jthread).
+ *     The test creates a new virtual thread.
+ *     Its method run() invokes the following methods:
+ *      - method A() that is blocked on a monitor
+ *      - method B() that is stopped at a breakpoint
+ *      - method C() that forces agent to send AssertionError exception to its own thread
+ *     All cases are using JVMTI StopThread to send an AssertionError object.
+ *
+ * @requires vm.continuations
+ * @run main/othervm/native -agentlib:StopThreadTest StopThreadTest
+ */
+
+import java.lang.AssertionError;
+
+public class StopThreadTest {
+    private static final String agentLib = "StopThreadTest";
+    static final int JVMTI_ERROR_NONE = 0;
+    static final int THREAD_NOT_SUSPENDED = 13;
+    static final int PASSED = 0;
+    static final int FAILED = 2;
+
+    static void log(String str) { System.out.println(str); }
+
+    static native void prepareAgent(Class taskClass, Object exceptionObject);
+    static native void suspendThread(Thread thread);
+    static native void resumeThread(Thread thread);
+    static native void ensureAtBreakpoint();
+    static native void notifyAtBreakpoint();
+    static native int  stopThread(Thread thread);
+
+    static int status = PASSED;
+
+    static void setFailed(String msg) {
+        log("\nFAILED: " + msg);
+        status = FAILED;
+    }
+
+    static void throwFailed(String msg) {
+        log("\nFAILED: " + msg);
+        throw new RuntimeException("StopThreadTest failed!");
+    }
+
+    public static void main(String args[]) {
+        run();
+        if (status == FAILED) {
+            throwFailed("StopThreadTest!");
+        }
+        log("\nStopThreadTest passed");
+    }
+
+    public static void run() {
+        TestTask testTask = new TestTask();
+        Thread testTaskThread = null;
+        AssertionError excObject = new AssertionError();
+        int retCode;
+
+        prepareAgent(TestTask.class, excObject);
+
+        log("\nMain #A: method A() must be blocked on entering a synchronized statement");
+        synchronized (TestTask.lock) {
+            testTaskThread = Thread.ofVirtual().name("TestTaskThread").start(testTask);
```

Comment:
Do we have other tests that are doing the equivalent testing on platform threads?

Reply:
Do you ask this question to check if we want to extend this test to provide coverage for platform threads as well?

The support of platform threads is much simpler. The JVMTI `StopThread` never returns the `THREAD_NOT_SUSPENDED` and `OPAQUE_FRAME` error codes for platform threads. So that this kind of testing is not needed for platform threads.

Reply:
For the JDI tests I added, I execute them in both modes, with the appropriate adjustments to account for the errors we except for virtual threads. We should be testing to make sure that StopThread works with platform threads under a variety of situations.

Reply:
Extending this test to cover platform threads does not look natural and is going to be a little ugly.
But I can extend it to provide coverage for BoundVirtualThread case
which is highjacking the platform thread implementation.
Would it help?
We should have pretty good coverage of the JVMTI `StopThread` for platform threads in `nsk.jvmti` test suite.
It includes:
  - `stopthrd006` and `stopthrd007`
  - a number of `scenarios/capability/CM01 `tests

Reply:
I've extended the test to cover platform threads as well.